library/md.c

```diff
@@ -575,15 +575,28 @@ int mbedtls_md_hmac_starts( mbedtls_md_context_t *ctx, const unsigned char *key,
     if( ( ret = mbedtls_md_info_starts( md_info, ctx->md_ctx ) ) != 0 )
         goto cleanup;
 
+    i++;    // Use i as flow control
+
     if( ( ret = mbedtls_md_info_update( md_info, ctx->md_ctx, ipad,
            mbedtls_md_info_block_size( md_info ) ) ) != 0 )
     {
         goto cleanup;
     }
 
+    i++;    // Use i as flow control now
+
 cleanup:
     mbedtls_platform_zeroize( sum, sizeof( sum ) );
 
+    if ( ret == 0 )
```

Comment:
The checking of doubles and illegal conditions can now be omitted by skipping this branching. It would be better to have a ret_fi -variable which is initially MBEDTLS_ERR_PLATFORM_FAULT_DETECTED and set to 0 when everything is Ok and returned in the end.